core/types/header_ext_test.go

```diff
@@ -0,0 +1,144 @@
+
+	header := headerWithNonZeroFields()
+	gotHashHex := header.Hash().Hex()
+	assert.Equal(t, "0x"+wantHashHex, gotHashHex, "Header.Hash()")
+}
+
+func TestHeaderJSON(t *testing.T) {
+	t.Parallel()
+
+	testHeaderEncodings(t, json.Marshal, json.Unmarshal)
+}
+
+func testHeaderEncodings(t *testing.T, encode func(any) ([]byte, error), decode func([]byte, any) error) []byte {
+	t.Helper()
+
+	input := headerWithNonZeroFields() // the Header carries the HeaderExtra so we can ignore it
+	b, err := encode(input)
+	require.NoError(t, err, "encode")
+
+	got := new(Header)
+	err = decode(b, got)
+	require.NoError(t, err, "decode")
+
+	want := headerWithNonZeroFields()
+	want.WithdrawalsHash = nil
+	assert.Equal(t, want, got)
+
+	return b
+}
+
+func Test_headerWithNonZeroFields(t *testing.T) {
+	t.Parallel()
+
+	header := headerWithNonZeroFields()
+	extra := GetHeaderExtra(header)
+	t.Run("Header", func(t *testing.T) { allExportedFieldsSet(t, header) })
+	t.Run("HeaderExtra", func(t *testing.T) { allExportedFieldsSet(t, extra) })
+}
+
+// headerWithNonZeroFields returns a [Header] with its extra payload [HeaderExtra]
+// set via [SetHeaderExtra], each with all fields set to non-zero values.
+//
+// NOTE: They can be used to demonstrate that RLP and JSON round-trip encoding
+// can recover all fields, but not that the encoded format is correct. This is
+// very important as the RLP encoding of a [Header] defines its hash.
+func headerWithNonZeroFields() *Header {
+	header := &ethtypes.Header{
+		ParentHash:       common.Hash{1},
+		UncleHash:        common.Hash{2},
+		Coinbase:         common.Address{3},
+		Root:             common.Hash{4},
+		TxHash:           common.Hash{5},
+		ReceiptHash:      common.Hash{6},
+		Bloom:            Bloom{7},
+		Difficulty:       big.NewInt(8),
+		Number:           big.NewInt(9),
+		GasLimit:         10,
+		GasUsed:          11,
+		Time:             12,
+		Extra:            []byte{13},
+		MixDigest:        common.Hash{14},
+		Nonce:            BlockNonce{15},
+		BaseFee:          big.NewInt(16),
+		WithdrawalsHash:  &common.Hash{17},
+		BlobGasUsed:      ptrTo(uint64(18)),
+		ExcessBlobGas:    ptrTo(uint64(19)),
+		ParentBeaconRoot: &common.Hash{20},
+	}
+	extra := &HeaderExtra{
+		ExtDataHash:    common.Hash{21},
+		ExtDataGasUsed: big.NewInt(22),
+		BlockGasCost:   big.NewInt(23),
+	}
+	SetHeaderExtra(header, extra)
+	return header
+}
+
+func allExportedFieldsSet[T interface {
+	ethtypes.Header | HeaderExtra
+}](t *testing.T, x *T) {
+	// We don't test for nil pointers because we're only confirming that
+	// test-input data is well-formed. A panic due to a dereference will be
+	// reported anyway.
+
+	v := reflect.ValueOf(*x)
+	for i := range v.Type().NumField() {
+		field := v.Type().Field(i)
+		if !field.IsExported() {
+			continue
+		}
+
+		t.Run(field.Name, func(t *testing.T) {
+			switch f := v.Field(i).Interface().(type) {
+			case common.Hash, common.Address, BlockNonce, Bloom, uint64:
+				assert.NotZero(t, f)
```

Comment:
(subjective rant) I very strongly dislike these "fancier" assert/require functions (including `IsType`) when there's a native language construct for achieving the same in an obviously correct manner. I had to go review the implementation of `NotZero()` to convince myself that it achieves exactly what we want.

I'm not entirely convinced that it's appropriate for `big.Int`.

Reply:
I get your points, but:

- one case was already using `assert.NotEmpty` it fell natural to use other assert.* functions
- assert.NotZero using `any` (not great) allows to combine switch cases together (great)
- your function `assertNonZero` felt like something that can be replaced with a lib-already-in-use-function

Regarding *big.Int, your line was `assert.NotEqual(t, 0, f.Cmp(big.NewInt(0)))` but what we really want is just for it to be non-nil, leaving it to `0` is fine in that case (not the zero value despite being `0` 😄)

For the require.IsType, yes it's not amazing (maybe one day we'll get generics with this lib, probably not 😄)  but it's an explicit one-liner, better than a switch case with generic 😉

Reply:
Reverted to using a local `assertNonZero` for cases in the switch except for []uint8